### PR TITLE
refactor: improve context type

### PR DESCRIPTION
**Pull request by iamchanii**
## About the changes

- Improve `FlagContext` type. so you can delete manual typing via `as`.

### Important files

- **src/FlagContext.ts**: it contains `IFlagContextValue` interface. 

## Discussion points

I'd like to do refactor `useVariant` typing, but I'd like to hear your opinions before that. how do you think If I refactor like this?:

```ts

type VariantPayload<TPayloadValue> =
  | JsonVariantPayload<TPayloadValue>
  | CsvVariantPayload
  | StringVariantPayload;

type IUseVariantResult<TPayloadValue> = {
  name: string;
  enabled: boolean;
  payload?: VariantPayload<TPayloadValue>;
};

const useVariant = <TPayloadValue extends {}>(
  name: string
): IUseVariantResult<TPayloadValue> => {
  // ...
}
```

[(TS Playground)](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgzgewHYDUCGAnAlmpwAKaIANgmgCYA8AKoSWeesQK4QB8UAvFAN4CwAKChRQkAFxQA5ACtESSQG5BwgG5oWECbSKkKTVkoEBfQYNHQAwnBXpsuAjoZdeykeE1SAxtcWu1GiThgbCQAc0MTATN3KABlYKww2xw8Ol1yZ34hN3EpIJDQ32z-VkCEsIjTAXMoZPs0hhoGvXVWDk5XAB9YOTrUxwomgcZW9i6oKxtMFId6CnH4gr7Z9MNoyCgASQBVOAhlgCUIOGZiYCG5kY12l2ykNABbD3zE8NcIe4AjYghyCU+EAgfrhDMIwMMAPwSZbNKjaS76dgRNYCTzIIJQZh7ZbOC7pRFQCAAD2AH3IcF4RjYAApXPcnmUCoIAJQSHbY6b2I4nM54hiI9ocLLCDAQYDMDBIKBIU7EKBoCm4ECVKICAD0aq2UosCAeYGQH2AYkEaKQGJ40seHjUdjwADkrQAaKDgy5QIzOLH7Tl4SgWrB-aXMB6fCAYd00yQAdwgxDRTwAtAAzYhoQrMlFYJNQamu9IQgB0NU4JaksmQkmZt2EGpyEAQ2bzDALJQgBYDXFLkhlIbDklcpsQPwLpFCueGLdG7fIzMEHtjeygWZzTYoheLXe8Kkr1agtfMDZdE9bne4kheYX72UHQLbo-Hl0nGlnxkJxEXy4f+aLMRLZ4v6a7vu7iHqu5BPqwp55OUhQDuit4jggY5gRBEAvkYQA)

### CODE REVIEW

**Comment by FredrikOseberg** [comment]
Thanks! We'll have a look

**Comment by FredrikOseberg** [comment]
[refactor: improve context type](https://trello.com/c/LMFBOw95/1177-refactor-improve-context-type)

**Comment by iamchanii** [comment]
> Do you mean typing `as` in end-user code or in the library?

No, I think end-users don't have to change some codes. because there are no significant changes in the existing test code. (exclude changes for type error)

This pr only affects internal code. for example, [this code](https://github.com/Unleash/proxy-client-react/blob/main/src/useVariant.ts#L8-L9) and [this code](https://github.com/Unleash/proxy-client-react/blob/main/src/useFlag.ts#L7-L8). so I think we can remove `any` type or `as UnleashClient` typing.

> ... though I'm not entirely sure how the payload there would work out.

You're right. and I'm not used to Unleash. if `json` variant provide stringified json, my suggestions must be rejected.

**Comment by thomasheartman** [comment]
> ~~No, I think end-users don't have to change some codes.~~ because there are no significant changes in the existing test code. (exclude changes for type error)
> 
> EDIT: Maybe need some changes if some end-user typed useVariant hook result manually and it's not incompatible. currently, useVariant's result type is any.

Got it! So this shouldn't break anyone's applications unless they have typed something explicitly themselves. That sounds alright ☺️ 


By the way, just for clarification: what's the motivation for this PR? Is it better type information as an end user?


> if json variant provide stringified json, my suggestions must be rejected.

The API returns stringified JSON, at least. I don't think this library or the base [proxy client for js](https://github.com/Unleash/unleash-proxy-client-js) does any parsing of the payload for you.  At least I didn't find any evidence of that when I went looking just now.

**Comment by iamchanii** [comment]
> what's the motivation for this PR? Is it better type information as an end user?

Recently, I tried to use variant and I found `useVariant`'s return type is `any`. Non-TypeScript users don't care about it. but, as a TypeScript user, `any` type is something should I avoid. if `useVariant` return `IVariant` type, it should be more helpful for an end-user. BTW, I had to log useVariant's result often.

File: src/FlagContext.ts [hunk 1]
```diff
@@ -1,5 +1,24 @@
 import React from 'react';
+import type { IContext, IVariant, UnleashClient } from 'unleash-proxy-client';
 
-const FlagContext = React.createContext(null);
+type eventArgs = [Function, any];
+
+export interface IFlagContextValue {
+  on(event: string, ...args: eventArgs): UnleashClient;
```
Comment (Tymek) [inline src/FlagContext.ts:7]: `eventArgs` can be inline (not exported type), and I think third argument is optional, isn't it?

**Review by Tymek — CHANGES_REQUESTED** [review]
I like this improvement a lot. One small request: type for `IFlagContextValue` can use `UnleashClient` more directly.

This gave me an idea to do some kind of CLI/generics to autocomplete feature names and variants from types. I hope it happens sometime down the line.

**Review by Tymek — APPROVED** [review]

**Comment by Tymek** [comment]
Thank you for this contribution. We will release it very soon :shipit:

**Pull request merged**